File: regions/system/library/csharp/health/HealthChecker.cs
namespace K1s0.System.Health;

public class HealthChecker
{
    private readonly List<IHealthCheck> _checks = new();

    public void Add(IHealthCheck check) => _checks.Add(check);

    public async Task<HealthResponse> RunAllAsync(CancellationToken ct = default)
    {
        var results = new Dictionary<string, CheckResult>();
        var worstStatus = HealthStatus.Healthy;

        foreach (var check in _checks)
        {
            var result = await check.CheckAsync(ct).ConfigureAwait(false);
            results[check.Name] = result;
            if (result.Status > worstStatus)
            {
                worstStatus = result.Status;
            }
        }

        return new HealthResponse(worstStatus, results, DateTimeOffset.UtcNow);
    }
}

File: regions/system/library/swift/health/Sources/K1s0Health/HealthChecker.swift
import Foundation

public actor HealthChecker {
    private var checks: [any HealthCheck] = []

    public init() {}

    public func add(_ check: some HealthCheck) {
        checks.append(check)
    }

    public func runAll() async -> HealthResponse {
        var results: [String: CheckResult] = [:]
        var overall = HealthStatus.healthy
        for c in checks {
            do {
                try await c.check()
                results[c.name] = CheckResult(status: .healthy)
            } catch {
                results[c.name] = CheckResult(status: .unhealthy, message: error.localizedDescription)
                overall = .unhealthy
            }
        }
        return HealthResponse(status: overall, checks: results, timestamp: Date())
    }
}
